### PR TITLE
Add Redis Session Support

### PR DESCRIPTION
**Pull request by jdolinski**
Pull Request for Issue # .

### Summary of Changes
Added support for Redis session handling.

### Testing Instructions
Modify the configuration.php file and include the following lines:
```
  public $session_redis_server_host = 'redis-server-url';
  public $session_redis_server_port = '6379';
```

### Expected result
Sessions are stored in Redis and slight performance improvement when used in conjunction with Redis Cache.

### Actual result
Session management is done by Redis and will work in a clustered environment without using database session handler.

### Documentation Changes Required
Modify configuration.php
```
  public $session_redis_server_host = 'redis-server-url';
  public $session_redis_server_port = '6379';
```

### CODE REVIEW

**Comment by jdolinski** [comment]
[Database session handling and Redis cache] (https://snag.gy/qiRb5m.jpg)
[Redis session handling and cache] (https://snag.gy/gi6Ih9.jpg)
[Redis session hits, looping 50 users across website pages](https://snag.gy/Ie4x9j.jpg)

**Review by mbabker — APPROVED** [review]

**Comment by mbabker** [comment]
I have tested this item :white_check_mark: successfully on a05a4e554f068498e47ad4b2b5e65aca64ddc7e0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15390">issues.joomla.org/tracker/joomla-cms/15390</a>.</sub>

**Comment by mbabker** [comment]
If you're a Mac user running Homebrew for PHP, this is actually pretty easy to test.

```sh
brew install redis
brew install phpxx-redis (replace xx with the PHP version you're using)
brew services start redis
```

Then get into global config, change the session handler (default settings are fine), save, and validate that things work using Redis as a session backend.

I'm comfortable merging based on my test, but it'd be nice if someone else could also do so.

File: libraries/joomla/session/storage/redis.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+		if (!self::isSupported())
+		{
+			throw new RuntimeException('Redis Extension is not available', 404);
+		}
+
+		$config = JFactory::getConfig();
+
+		$this->_server = array(
+			'host' => $config->get('session_redis_server_host', 'localhost'),
+			'port' => $config->get('session_redis_server_port', 11211),
```
Comment (alikon) [inline libraries/joomla/session/storage/redis.php]: why 11211 ?
shouldn't be 6379

**Review by alikon — CHANGES_REQUESTED** [review]

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on a05a4e554f068498e47ad4b2b5e65aca64ddc7e0<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/15390">issues.joomla.org/tracker/joomla-cms/15390</a>.</sub>


but we still need to manage situation like mod_status that directly access to the #__session table https://github.com/joomla/joomla-cms/blob/staging/administrator/modules/mod_status/mod_status.php#L33

File: libraries/joomla/session/storage/redis.php (hunk 1 above)
Comment (mbabker) [reply libraries/joomla/session/storage/redis.php]: Probably bad copy/paste, I fixed it.

**Comment by mbabker** [comment]
> but we still need to manage situation like mod_status that directly access to the #__session table

It's fine for now because of the app code that arbitrarily writes stuff to the session table even if the database session handler isn't in use.

**Pull request merged**